langchain/agents/agent.py

```diff
@@ -399,3 +446,77 @@ def _call(self, inputs: Dict[str, str]) -> Dict[str, Any]:
             self.early_stopping_method, intermediate_steps, **inputs
         )
         return self._return(output, intermediate_steps)
+
+    async def _acall(self, inputs: Dict[str, str]) -> Dict[str, str]:
+        """Run text through and get agent response."""
+        # Make sure that every tool is asynchronous (a coroutine)
+        for tool in self.tools:
+            if not asyncio.iscoroutinefunction(tool.coroutine):
+                raise ValueError(
+                    "The coroutine for the tool must be a coroutine function."
+                )
+
+        # Do any preparation necessary when receiving a new input.
+        self.agent.prepare_for_new_call()
+        # Construct a mapping of tool name to tool for easy lookup
+        name_to_tool_map = {tool.name: tool for tool in self.tools}
+        # We construct a mapping from each tool to a color, used for logging.
+        color_mapping = get_color_mapping(
+            [tool.name for tool in self.tools], excluded_colors=["green"]
+        )
+        intermediate_steps: List[Tuple[AgentAction, str]] = []
+        # Let's start tracking the iterations the agent has gone through
+        iterations = 0
+        # We now enter the agent loop (until it returns something).
+        while self._should_continue(iterations):
+            # Call the LLM to see what to do.
+            output = await self.agent.aplan(intermediate_steps, **inputs)
+            # If the tool chosen is the finishing tool, then we end and return.
+            if isinstance(output, AgentFinish):
+                return self._return(output, intermediate_steps)
+
+            # Otherwise we lookup the tool
+            if output.tool in name_to_tool_map:
+                tool = name_to_tool_map[output.tool]
+                self.callback_manager.on_tool_start(
+                    {"name": str(tool.func)[:60] + "..."},
+                    output,
+                    verbose=self.verbose,
+                )
+                try:
+                    # We then call the tool on the tool input to get an observation
+                    observation = (
+                        await tool.coroutine(output.tool_input)
+                        if tool.coroutine
+                        else tool.func(output.tool_input)
```

Comment:
This should probably be called with one of the asyncio methods for executing sync functions in another thread. If left like this it blocks the event loop for the duration of the tool call

Reply:
```py
asyncio.get_event_loop().run_in_executor(None, tool.func, output.tool_input)
```

Reply:
Yep, great point